File: Commons/Action/src/main/java/com/greentree/commons/action/observable/ObjectObservable.java
package com.greentree.commons.action.observable;

import java.util.function.Consumer;

import com.greentree.commons.action.ListenerCloser;

public interface ObjectObservable<T> extends RunObservable {

	ObjectObservable<?> NULL = new ObjectObservable<>() {
		private static final long serialVersionUID = 1L;

		@Override
		public ListenerCloser addListener(Consumer<? super Object> listener) {
			return ListenerCloser.EMPTY;
		}

		@Override
		public int listenerSize() {
			return 0;
		}

	};

	@SuppressWarnings("unchecked")
	static <T> ObjectObservable<T> getNull() {
		return (ObjectObservable<T>) NULL;
	}

	ListenerCloser addListener(Consumer<? super T> listener);

	@Override
	default ListenerCloser addListener(Runnable listener) {
		return addListener(t -> listener.run());
	}

}

File: Commons/Action/src/main/java/com/greentree/commons/action/observable/ObjectObservable.kt
package com.greentree.commons.action.observable

import com.greentree.commons.action.ListenerCloser
import java.util.function.Consumer

interface ObjectObservable<T> : RunObservable {

	override fun addListener(listener: Runnable) = addListener { _ -> listener.run() }

	fun addListener(listener: (T) -> Unit): ListenerCloser

	data object Empty : ObjectObservable<Any?> {

		fun <T> instance() = Empty as ObjectObservable<T>

		override val listenerCount: Int
			get() = 0

		override fun addListener(listener: (Any?) -> Unit): ListenerCloser {
			return ListenerCloser.Empty
		}
	}
}

@Deprecated("", ReplaceWith("addListener { listener.accept(it) }"))
fun <T> ObjectObservable<T>.addListener(listener: Consumer<in T>): ListenerCloser = addListener { listener.accept(it) }